### PR TITLE
fix(PerpV2BasisTradingModule): bytecode size limit bug

### PR DESCRIPTION
**Pull request by 0xSachinK**
## Bytecode Sizes
Limit: 24,567 bytes

### Master Branch
PerpV2LeverageModule:  23,970 bytes
PerpV2BasisTradingModule:  27,499 bytes

### Current Branch
PerpV2LeverageModule:  23,970 bytes
PerpV2LeverageModuleV2:  20,680 bytes
PerpV2BasisTradingModule:  24,339 bytes

## ChangeLog
- Create `V2` versions of `Position` library, `PerpV2` library, `PerpV2LeverageModule`, named `PositionV2`, `PerpV2LibraryV2` and `PerpV2LeverageModuleV2` respectively.
- `PositionV2` is a linkable library and contains external/public versions of functions available in `Position`.
- `PerpV2LibraryV2` contains `executeTrade` and `simulateTrade` generic functions, moved over from `PerpV2LeverageModule`.
- New `PerpV2Positions` library contains generic getter functions, `getNetQuoteBalance`, `getPositionUnitInfo` and `getPositionNotionalInfo`, moved over from `PerpV2LeverageModule`.
- `ModuleBaseV2` uses the new `PositionV2` library contract.
- `PerpV2LeverageModuleV2` extends `ModuleBaseV2` and uses the new libraries, and thus has a smaller bytecode size than `PerpV2LeverageModule`.
- `PerpV2BasisTradingModule` now extends `PerpV2LeverageModuleV2` instead of the V1 version.

## Diffs
- [Position <> PositionV2](https://www.diffchecker.com/RasBVLX8)
- [ModuleBase <> ModuleBaseV2](https://www.diffchecker.com/yHKd34YF)
- [PerpV2 <> PerpV2LibraryV2](https://www.diffchecker.com/PtkRmmpe)
- [PerpV2LeverageModule <> PerpV2LeverageModuleV2](https://www.diffchecker.com/Np5GXXTM)
- New Library: [PerpV2Positions](https://github.com/SetProtocol/set-protocol-v2/blob/sachin/update-allowUnlimitedContractSize/contracts/protocol/integration/lib/PerpV2Positions.sol)
- [PerpV2BasisTradingModule master branch <> PerpV2BasisTradingModule this branch](https://www.diffchecker.com/WowYW0gj)

## Todo
- [x] Fix hardhat config file
- [x] Fix existing flaky tests
- [x] Specify security aspects of PositionV2
- [x] Move formatAdjustments to a new library contract to decrease it further (Don't pass in components as a parameter).

## Tests
- [x] PositionV2 library
- [x] ModuleBaseV2 library
- [x] PerpV2LibraryV2 library
- [x] PerpV2Positions library
- [x] PerpV2LeverageModuleV2
- [x] PerpV2BasisTradingModule

## Security

### CODE REVIEW

File: contracts/interfaces/external/perp-v2/IMarketRegistry.sol [hunk 1]
```diff
@@ -23,50 +23,50 @@ interface IMarketRegistry {
...
 
-    function getMarketInfo(address baseToken) external view returns (MarketInfo memory);
+    // function getMarketInfo(address baseToken) external view returns (MarketInfo memory);
 
     function getQuoteToken() external view returns (address);
 
     function getUniswapV3Factory() external view returns (address);
 
-    function getMaxOrdersPerMarket() external view returns (uint8);
+    // function getMaxOrdersPerMarket() external view returns (uint8);
```
Comment (bweick) [inline contracts/interfaces/external/perp-v2/IMarketRegistry.sol:69]: Let's delete all these commented out fxns. Unless there's still something you're checking

File: contracts/protocol/integration/lib/PerpV2.sol [hunk 2]
```diff
@@ -198,7 +207,7 @@ library PerpV2 {
         IClearingHouse _clearingHouse,
         IClearingHouse.OpenPositionParams memory _params
     )
-        external
+        public
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2.sol:210]: Is there a reason to make this `public`?

File: contracts/protocol/modules/PerpV2LeverageModule.sol [hunk 3]
```diff
@@ -881,12 +832,12 @@ contract PerpV2LeverageModule is ModuleBase, ReentrancyGuard, Ownable, SetTokenA
...
 
-        _setToken.editExternalPosition(
-            address(collateralToken),
-            address(this),
-            _calculateExternalPositionUnit(_setToken),
-            ""
-        );
+        // _setToken.editExternalPosition(
+        //     address(collateralToken),
+        //     address(this),
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModule.sol:837]: pls delet

File: contracts/protocol/modules/PerpV2LeverageModule.sol [hunk 4]
```diff
@@ -937,71 +888,16 @@ contract PerpV2LeverageModule is ModuleBase, ReentrancyGuard, Ownable, SetTokenA
...
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModule.sol:893]: pls delet

File: contracts/protocol/modules/PerpV2LeverageModule.sol [hunk 5]
```diff
@@ -1170,12 +1066,12 @@ contract PerpV2LeverageModule is ModuleBase, ReentrancyGuard, Ownable, SetTokenA
...
      */
-    function _calculateAMMSpotPrice(address _baseToken) internal view returns (uint256 price) {
-        address pool = perpMarketRegistry.getPool(_baseToken);
-        (uint160 sqrtPriceX96, , , , , , ) = IUniswapV3Pool(pool).slot0();
-        uint256 priceX96 = sqrtPriceX96.formatSqrtPriceX96ToPriceX96();
-        return priceX96.formatX96ToX10_18();
-    }
+    // function _calculateAMMSpotPrice(address _baseToken) internal view returns (uint256 price) {
+    //     address pool = perpMarketRegistry.getPool(_baseToken);
+    //     (uint160 sqrtPriceX96, , , , , , ) = IUniswapV3Pool(pool).slot0();
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModule.sol:1071]: pls delet

File: contracts/protocol/modules/PerpV2LeverageModule.sol [hunk 6]
```diff
@@ -1184,50 +1080,27 @@ contract PerpV2LeverageModule is ModuleBase, ReentrancyGuard, Ownable, SetTokenA
...
+    //         _setToken, 
+    //         positions[_setToken], 
+    //         perpAccountBalance
+    //     );
+    //     uint256 positionLength = positionInfo.length;
+    //     int256 totalPositionValue = 0;
+
+    //     for (uint i = 0; i < positionLength; i++ ) {
+    //         int256 spotPrice = _calculateAMMSpotPrice(positionInfo[i].baseToken).toInt256();
+    //         totalPositionValue = totalPositionValue.add(
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModule.sol:1094]: pls delet

File: hardhat.config.ts [hunk 7]
```diff
@@ -68,6 +69,10 @@ const config: HardhatUserConfig = {
     target: "ethers-v5",
     externalArtifacts: ["external/**/*.json"],
   },
+  // contractSizer: {
+  //   // alphaSort: true,
```
Comment (bweick) [inline hardhat.config.ts:73]: delet

File: contracts/protocol/modules/PerpV2LeverageModule.sol [hunk 8]
```diff
@@ -776,7 +723,11 @@ contract PerpV2LeverageModule is ModuleBase, ReentrancyGuard, Ownable, SetTokenA
         // and variable may refer to the value which will be redeemed.
         int256 accountValueIssued = _calculatePartialAccountValuePositionUnit(_setToken).preciseMul(setTokenQuantityInt);
 
-        PositionNotionalInfo[] memory positionInfo = getPositionNotionalInfo(_setToken);
+        PerpV2Positions.PositionNotionalInfo[] memory positionInfo = PerpV2Positions.getPositionNotionalInfo(
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModule.sol:726]: There a reason to use this and not the `public` function defined above?

File: contracts/protocol/integration/lib/PerpV2Positions.sol [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/math/SignedSafeMath.sol";
+import { IAccountBalance } from "../../../interfaces/external/perp-v2/IAccountBalance.sol";
+import { ISetToken } from "../../../interfaces/ISetToken.sol";
+import { PreciseUnitMath } from "../../../lib/PreciseUnitMath.sol";
+import { SafeCast } from "@openzeppelin/contracts/utils/SafeCast.sol";
+import { UnitConversionUtils } from "../../../lib/UnitConversionUtils.sol";
+
+// Todo: Typechain doesn't create types when parameters are passed by storage. Although saves only 68 bytes.
```
Comment (0xSachinK) [inline contracts/protocol/integration/lib/PerpV2Positions.sol:29]: @cgewecke Need your help with this.

Comment (cgewecke) [reply contracts/protocol/integration/lib/PerpV2Positions.sol:29]: Weird. Do you have to pass by storage if you're well under the limit?

Comment (0xSachinK) [reply contracts/protocol/integration/lib/PerpV2Positions.sol:29]: No. These are getter functions, so it works if I pass by memory.

File: test/protocol/modules/perpV2LeverageModuleV2.spec.ts [hunk 10]
```diff
@@ -0,0 +1,6062 @@
+import "module-alias/register";
```
Comment (bweick) [inline test/protocol/modules/perpV2LeverageModuleV2.spec.ts:1]: Can we also add tests for the new libraries? Especially want to make sure nothing wonky can happen where someone can call the library and it actually updates state. Thinking specifically of the `PositionV2.sol`

File: contracts/protocol/lib/PositionV2.sol [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+import { SafeCast } from "@openzeppelin/contracts/utils/SafeCast.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { SignedSafeMath } from "@openzeppelin/contracts/math/SignedSafeMath.sol";
+
+import { ISetToken } from "../../interfaces/ISetToken.sol";
+import { PreciseUnitMath } from "../../lib/PreciseUnitMath.sol";
+
+
+/**
+ * @title Position
```
Comment (bweick) [inline contracts/protocol/lib/PositionV2.sol:32]: `PositionV2`

File: contracts/protocol/lib/PositionV2.sol [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+ * @author Set Protocol
+ *
+ * Collection of helper functions for handling and updating SetToken Positions
+ *
+ * CHANGELOG:
+ *  - Updated editExternalPosition to work when no external position is associated with module
```
Comment (bweick) [inline contracts/protocol/lib/PositionV2.sol:38]: Can get rid of this change log but would add that the V2 version was built to be able to link it

File: contracts/protocol/lib/ModuleBaseV2.sol [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+import { Invoke } from "./Invoke.sol";
+import { PositionV2 } from "./PositionV2.sol";
+import { PreciseUnitMath } from "../../lib/PreciseUnitMath.sol";
+import { ResourceIdentifier } from "./ResourceIdentifier.sol";
+import { SafeCast } from "@openzeppelin/contracts/utils/SafeCast.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { SignedSafeMath } from "@openzeppelin/contracts/math/SignedSafeMath.sol";
+
+/**
+ * @title ModuleBase
```
Comment (bweick) [inline contracts/protocol/lib/ModuleBaseV2.sol:37]: Title

File: contracts/protocol/integration/lib/PerpV2LibraryV2.sol [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IClearingHouse } from "../../../interfaces/external/perp-v2/IClearingHouse.sol";
+import { IVault } from "../../../interfaces/external/perp-v2/IVault.sol";
+import { IQuoter } from "../../../interfaces/external/perp-v2/IQuoter.sol";
+import { ISetToken } from "../../../interfaces/ISetToken.sol";
+import { PreciseUnitMath } from "../../../lib/PreciseUnitMath.sol";
+
+/**
+ * @title PerpV2
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2LibraryV2.sol:30]: Title

File: contracts/protocol/integration/lib/PerpV2LibraryV2.sol [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+ * @author Set Protocol
+ *
+ * Collection of helper functions for interacting with PerpV2 integrations.
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2LibraryV2.sol:33]: Add changelog. What was added from last version and mention we did this to make it linkable

File: contracts/protocol/modules/PerpV2LeverageModuleV2.sol [hunk 16]
```diff
@@ -0,0 +1,1134 @@
...
+ * represented as a positive equity external position whose value is the net Perp account value denominated in the collateral token
+ * deposited into the Perp Protocol. This module only allows Perp positions to be collateralized by one asset, USDC, set on deployment of
+ * this contract (see collateralToken) however it can take positions simultaneously in multiple base assets.
+ *
+ * Upon issuance and redemption positions are not EXACTLY replicated like for other position types since a trade is necessary to enter/exit
+ * the position on behalf of the issuer/redeemer. Any cost of entering/exiting the position (slippage) is carried by the issuer/redeemer.
+ * Any pending funding costs or PnL is carried by the current token holders. To be used safely this module MUST issue using the
+ * SlippageIssuanceModule or else issue and redeem transaction could be sandwich attacked.
+ *
+ * NOTE: The external position unit is only updated on an as-needed basis during issuance/redemption. It does not reflect the current
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2LeverageModuleV2.sol:63]: Maybe add a small summary of changes from V1

File: contracts/protocol/integration/lib/PerpV2Positions.sol [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    struct PositionUnitInfo {
+        address baseToken;              // Virtual token minted by the Perp protocol
+        int256 baseUnit;                // Base position unit. When negative, position is short
+        int256 quoteUnit;               // vUSDC "debt" position unit. When positive, position is short
+    }
+
+    /**
+     * @dev Retrieves net quote balance of all open positions.
+     *
+     * @param _setToken             Instance of SetToken
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2Positions.sol:57]: params

File: contracts/protocol/integration/lib/PerpV2Positions.sol [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+            netQuoteBalance = netQuoteBalance.add(
+                _perpAccountBalance.getQuote(address(_setToken), _baseTokens[i])
+            );
+        }
+    }
+
+    /**
+     * @dev Returns a PositionUnitNotionalInfo array representing all positions open for the SetToken.
+     *
+     * @param _setToken         Instance of SetToken
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2Positions.sol:80]: params

File: contracts/protocol/integration/lib/PerpV2Positions.sol [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+            });
+        }
+
+        return positionInfo;
+    }
+    
+    /**
+     * @dev Returns a PerpV2Positions.PositionUnitInfo array representing all positions open for the SetToken.
+     *
+     * @param _setToken         Instance of SetToken
```
Comment (bweick) [inline contracts/protocol/integration/lib/PerpV2Positions.sol:121]: params

File: contracts/protocol/modules/DebtIssuanceModuleV2.sol [hunk 20]
```diff
@@ -350,24 +350,19 @@ contract DebtIssuanceModuleV2 is DebtIssuanceModule {
         returns (address[] memory, uint256[] memory, uint256[] memory)
     {
         address[] memory components = _setToken.getComponents();
-        uint256 componentsLength = components.length;
-
-        uint256[] memory equityUnits = new uint256[](componentsLength);
-        uint256[] memory debtUnits = new uint256[](componentsLength);
-
-        uint256 totalSupply = _setToken.totalSupply();
+        uint256[] memory equityUnits = new uint256[](components.length);
```
[the remark this answers is not in the document]
Comment (0xSachinK) [reply contracts/protocol/modules/DebtIssuanceModuleV2.sol]: Removed.

File: test/protocol/modules/perpV2LeverageModuleV2.spec.ts (hunk 10 above)
Comment (0xSachinK) [reply test/protocol/modules/perpV2LeverageModuleV2.spec.ts:1]: Added tests in [f61d144](https://github.com/SetProtocol/set-protocol-v2/pull/200/commits/f61d14480d9e3e4d50a73b9697733d5ca1700da9) but these are an exact copy of the existing `Positions` library tests. But that is an interesting point, and we should consider the security aspects of using linked library contracts to modify the state of SetTokens.

File: hardhat.config.ts (hunk 7 above)
Comment (0xSachinK) [reply hardhat.config.ts]: Fixed.

File: hardhat.config.ts [hunk 21]
```diff
@@ -69,10 +69,9 @@ const config: HardhatUserConfig = {
     target: "ethers-v5",
     externalArtifacts: ["external/**/*.json"],
   },
-  // contractSizer: {
-  //   // alphaSort: true,
-  //   runOnCompile: true,
-  // },
+  contractSizer: {
```
[the remark this answers is not in the document]
Comment (0xSachinK) [reply hardhat.config.ts:73]: Chris on slack: "Wrap it in a `ts-ignore` section".

File: contracts/protocol-viewers/PerpV2LeverageModuleViewer.sol [hunk 22]
```diff
@@ -30,6 +30,7 @@ import { IClearingHouseConfig } from "../interfaces/external/perp-v2/IClearingHo
 import { IIndexPrice } from "../interfaces/external/perp-v2/IIndexPrice.sol";
 import { IPerpV2LeverageModule } from "../interfaces/IPerpV2LeverageModule.sol";
 import { ISetToken } from "../interfaces/ISetToken.sol";
+import { PerpV2Positions } from "../protocol/integration/lib/PerpV2Positions.sol";
```
[the remark this answers is not in the document]
Comment (0xSachinK) [reply contracts/protocol-viewers/PerpV2LeverageModuleViewer.sol:33]: We can't use the current viewer as it has the module from which it is fetching data as a state variable set to `PerpV2LeverageModule` in the constructor. We would need to write a new V2 viewer contract, which accepts the module to fetch data from in every function.

File: contracts/protocol/modules/PerpV2BasisTradingModule.sol [hunk 23]
```diff
@@ -391,28 +391,27 @@ contract PerpV2BasisTradingModule is PerpV2LeverageModuleV2 {
...
-            int256 newExternalPositionUnit = _executePositionTrades(_setToken, _setTokenQuantity, false, true);
-
             // Calculate performance fee unit
             // Performance fee unit = (Tracked settled funding * Performance fee) / Set total supply
             uint256 performanceFeeUnit = _getUpdatedSettledFunding(_setToken)
                 .preciseDiv(_setToken.totalSupply())
                 .preciseMulCeil(_performanceFeePercentage(_setToken))
                 .fromPreciseUnitToDecimals(collateralDecimals);
 
+            int256 partialNewExternalPositionUnit = _executePositionTrades(_setToken, _setTokenQuantity, false, true);
```
Comment (bweick) [inline contracts/protocol/modules/PerpV2BasisTradingModule.sol:404]: partial is kinda hard to understand in this context. If you want to communicate it's the unit without fees I'd be more explicit about it. `newExternalUnitWithoutFees` or something like that

**Review by bweick — APPROVED** [review]

**Pull request merged**